extensions/amp-next-page/0.1/amp-next-page.js

```diff
@@ -70,24 +76,128 @@ export class AmpNextPage extends AMP.BaseElement {
+              // Resolve this promise with an empty array anyway so we can use
+              // the inline/src config as a fallback.
+              return [];
+            });
+      }
+
+      const inlineConfig = this.getInlineConfig_();
+
       if (src) {
-        return this.fetchConfig_().then(
-            config => this.register_(service, config, separator),
+        configPromise = this.fetchConfig_().catch(
             error => user().error(TAG, 'error fetching config', error));
       } else {
-        const scriptElements = childElementsByTag(element, 'SCRIPT');
-        user().assert(scriptElements.length === 1,
-            `${TAG} should contain only one <script> child, or a URL specified `
-            + 'in [src]');
-        const scriptElement = scriptElements[0];
-        user().assert(isJsonScriptTag(scriptElement),
-            `${TAG} config should ` +
-            'be inside a <script> tag with type="application/json"');
-        const configJson = tryParseJson(scriptElement.textContent, error => {
-          user().error(TAG, 'failed to parse config', error);
-        });
-        this.register_(service, configJson, separator);
+        configPromise = Promise.resolve(inlineConfig);
       }
+
+      user().assert(inlineConfig || src || type,
+          `${TAG} should contain a <script> child, a URL specified in [src], `
+          + 'or a [type]');
+
+      return Promise.all([configPromise, pagesPromise]).then(values => {
+        const config = values[0] || {};
+        const pages = values[1] || [];
+        config.pages = pages.concat(config.pages || []);
+        this.register_(service, config, separator);
+      });
+    });
+  }
+
+  /**
+   * Reads the inline config from the element.
+   * @return {?*} Config JSON object, or null if no inline config specified.
+   * @private
+   */
+  getInlineConfig_() {
+    const scriptElements = childElementsByTag(this.element, 'SCRIPT');
+    if (!scriptElements.length) {
+      return null;
+    }
+    user().assert(scriptElements.length === 1,
+        `${TAG} should contain at most one <script> child`);
+    const scriptElement = scriptElements[0];
+    user().assert(isJsonScriptTag(scriptElement),
+        `${TAG} config should ` +
+        'be inside a <script> tag with type="application/json"');
+    return tryParseJson(scriptElement.textContent, error => {
+      user().error(TAG, 'failed to parse config', error);
+    });
+  }
+
+  /**
+   * Fetches content recommendations from AdSense and returns a list of {@link
+   * AmpNextPageItem}.
+   * @param {string} client AdSense publisher code.
+   * @param {string} slot AdSense Matched Content ad slotname.
+   * @param {boolean} personalized {@code true} if the request should be
+   *     personalized (with cookies).
+   * @return {!Promise<Array<./config.AmpNextPageItem>>} List of recommended
+   *     pages.
+   * @private
+   */
+  fetchAdSensePages_(client, slot, personalized) {
```

Comment:
I'd prefer to have AdSense server to return the format we defined, instead of having this AdSense specific logic in this generic extension.

It's not scalable if later we have all those content recommendation platform to add a piece of code here.

Reply:
We won't be able to get the AdSense server to return this format, XML is the only option for now.